cloudflare/terraform-provider-cloudflare
# Add terraform resource and data source for Infrastructure Access Target

Adds the resource and data source that we need to CRUDL infrastructure access target resources using the `cloudflare/cloudflare` terraform provider. 
Our OpenAPI schema is published in `cloudflare/api-schemas` viewable in [this commit](https://github.com/cloudflare/api-schemas/commit/07d60aae32323373b4a35ff1b8c2246ad927ef9e)
Our API docs are accessible in production at https://developers.cloudflare.com/api/operations/infra-targets-list

This PR is dependent on `cloudflare-go` client changes found in this PR -> https://github.com/cloudflare/cloudflare-go/pull/3184

New changes:
- New resource `cloudflare_infrastructure_access_target`
- New data source `cloudflare_infrastructure_access_targets`
- Added schema and acceptance tests for both the resource and data source

Manual test cases covered for resources:
- Creating a resource with ipv4 configuration, no ipv6. 
- Updating the resource to have a different ipv4. Then updating once more to add an ipv6 configuration.
- Destroying the resource and cleaning up terraform state once deleted

Manual test cases covered for data sources:
- Created data source to read all targets
- Added new target and refreshed which updated data source
- Tested creating data source with query parameters

Manual testing setup:
1. Checkout `sdadireddy/infrastructure-access` branch in `cloudflare-go`
2. Replace the `cloudflare-go` import in In `terraform-provider-cloudflare` by adding `replace github.com/cloudflare/cloudflare-go => ../cloudflare-go`
Replace the `cloudflare-go` in `go.mod`
3. In `terraform-provider-cloudflare`, run `make build-dev` build a new binary with the new changes
4. Create a test repository and configure your `~/.terraformrc` to add a development override that points terraform to the directory with the compiled binary (should just be the root of `terraform-provider-cloudflare`
5. Create a `main.tf` file to create a new resource as such:
```
terraform {
    required_providers {
        cloudflare = {
            source = "cloudflare/cloudflare"
        }
    }
}

provider "cloudflare" {
    api_token = "" // create api token in dash.cloudflare
}

resource "cloudflare_infrastructure_access_target" "sai-tftest-target" {
    account_id = "" // account id can be found in dash.cloudflare
    hostname   = "sai-tf-test"
    ip {
	ipv4 = {
              ip_addr = "250.26.29.250"
              virtual_network_id = "01920a8c-dc14-7bb2-b67b-14c858494a54"
	}
    }
}
```
7. Run `terraform init` and `TF_LOG=DEBUG terraform plan` to see the proposed new changes
8. Run `TF_LOG=DEBUG  terraform apply` and create the target
9. Query the target via [public api](https://developers.cloudflare.com/api/operations/infra-targets-post)
